splinepy/helpme/fit.py

```diff
@@ -0,0 +1,738 @@
+import numpy as _np
+
+from splinepy import settings as _settings
+from splinepy.utils import log as _log
+from splinepy.utils.data import has_scipy as _has_scipy
+from splinepy.utils.data import make_matrix as _make_matrix
+
+if _has_scipy:
+    from scipy.sparse.linalg import spsolve as _spsolve
+
+    # from scipy.sparse.linalg import lsqr as _lsqr
+
+
+def parametrize_curve(points, n_points, centripetal):
+    """
+    Parametrizes given points to be used in interpolation/approximation.
+    Values are then used to build adequate knot vectors.
+
+    Parameters
+    ----------
+    points: (m + 1 x dim) array
+        points to be interpolated/approximated
+
+    n_points: int
+        number of points
+
+    centripetal: bool
+        if True -> centripetal parametrization will be used
+
+    Returns
+    -------
+    u_k: (n x 1) array
+        array containing parametrization values
+    """
+    u_k = _np.empty((n_points, 1))
+
+    chord_lengths = _np.linalg.norm(_np.diff(points, axis=0), axis=1)
+
+    if centripetal:
+        chord_lengths = _np.sqrt(chord_lengths)
+
+    total_chord_length = _np.sum(chord_lengths)
+
+    chord_lengths = chord_lengths / total_chord_length
+    u_k[0] = 0.0
+    u_k[-1] = 1.0
+
+    u_k[1:-1] = _np.cumsum(chord_lengths[:-1]).reshape(-1, 1)
+    return u_k.reshape(-1, 1)
+
+
+def parametrize_surface(points, size, centripetal):
+    """
+    Parametrizes the given points to be used in interpolation/approximation.
+    Values are then used to build adequate knot vectors.
+
+    Parameters
+    ----------
+    points: (m + 1 x dim) array
+        points to be interpolated/approximated
+
+    size: (2,) array-like
+        number of points in u and v direction
+
+    centripetal: bool
+        if True -> centripetal parametrization will be used
+
+    Returns
+    -------
+    [u_k, v_l]: list
+        list containing parametrization values in both directions
+    """
+
+    from splinepy.helpme.multi_index import MultiIndex
+
+    mi = MultiIndex(size)
+
+    u_k = _np.zeros((size[0], 1))
+
+    for i in range(size[1]):
+        u_k += parametrize_curve(
+            # points=points[i * size[0] : (i + 1) * size[0]],
+            points=points[mi[i, :]],
+            n_points=size[0],
+            centripetal=centripetal,
+        )
+    u_k /= size[1]
+
+    v_l = _np.zeros((size[1], 1))
+
+    for k in range(size[0]):
+        v_l += parametrize_curve(
+            # points=points[range(k, size[0] * size[1], size[0])],
+            points=points[mi[:, k]],
+            n_points=size[1],
+            centripetal=centripetal,
+        )
+    v_l /= size[0]
+
+    return [u_k, v_l]
```

Comment:
```suggestion
    # Loop over all dimensions and append para_coords to list
    para_coords = []
    n_dimensions = len(size)
    for n_points in size:
        entry_indices = [slice(None,None,None) for _ in range(n_dimensions)]
        for i in range(n_points):
          u_k += parametrize_curve(
              points=points[mi[entry_indices]],
              n_points=n_points,
              centripetal=centripetal,
          )
      u_k /= size[1]
      
      para_coords.append(u_k)

    return para_coords
```


I am not sure if this is going to work. But I like this concept more, as it will also work for n-dimensional fitting... Please try this and give me feedback

Reply:
I kind of implemented it this way - but i still had to distinguish between curve and surface, since the indexing is different (additional dimension) and I didn't find a nice way to circumvent this problem.